src/engine/sparqlExpressions/NumericBinaryExpressions.cpp

```diff
@@ -68,9 +66,198 @@ inline auto andLambda = [](TernaryBool a, TernaryBool b) {
+  if (!leftChildExprs.has_value() || !rightChildExprs.has_value()) {
+    // If we have two children that yield no PrefilterExpressions, then there
+    // is also no PrefilterExpression over the AND (&&) conjunction available.
+    if (!leftChildExprs.has_value() && !rightChildExprs.has_value()) {
+      return std::nullopt;
+    }
+    // Given that only one of the children returned a PrefilterExpression
+    // vector and the pre-filter construction logic of AND (&&), we can just
+    // return the corresponding (non-empty) vector.
+    //
+    // EXAMPLE why this works:
+    // Assume that left is - {(>=5, ?x), (<=6, ?y)}
+    // and right is - std::nullopt ({undefinable prefilter})
+    // Remark: left represents the expression ?x >= 5 && ?y <= 6
+    //
+    // For this AND conjunction {(>=5, ?x), (<=6, ?y)} && {undef.
+    // prefilter}, we can at least (partly) pre-define the plausible true
+    // evaluation range with the left child. This is because the respective
+    // expressions contained in left child must always yield true
+    // for making this AND conjunction expression true.
+    return leftChildExprs.has_value() ? std::move(leftChildExprs.value())
+                                      : std::move(rightChildExprs.value());
+  }
+  // Merge the PrefilterExpression vectors from the left and right child.
+  // Remark: The vectors contain std::pairs, sorted by their respective
+  // variable to the PrefilterExpression.
+  auto& expressionsLeft = leftChildExprs.value();
+  auto& expressionsRight = rightChildExprs.value();
+  auto itLeft = expressionsLeft.begin();
+  auto itRight = expressionsRight.begin();
+  std::vector<PrefilterExprVariablePair> resPairs;
+  while (itLeft != expressionsLeft.end() && itRight != expressionsRight.end()) {
+    auto& [exprLeft, varLeft] = *itLeft;
+    auto& [exprRight, varRight] = *itRight;
+    // For our pre-filtering logic over index scans, we need exactly one
+    // PrefilterExpression for each of the variables. Thus if the left and right
+    // child contain a PrefilterExpression w.r.t. the same variable, combine
+    // them here over a (new) prefilter AndExpression.
+    if (varLeft == varRight) {
+      resPairs.emplace_back(std::make_unique<BinaryPrefilterExpr>(
+                                std::move(exprLeft), std::move(exprRight)),
+                            varLeft);
+      ++itLeft;
+      ++itRight;
+    } else if (varLeft < varRight) {
+      resPairs.emplace_back(std::move(*itLeft));
+      ++itLeft;
+    } else {
+      resPairs.emplace_back(std::move(*itRight));
+      ++itRight;
+    }
+  }
+  std::ranges::move(itLeft, expressionsLeft.end(),
+                    std::back_inserter(resPairs));
+  std::ranges::move(itRight, expressionsRight.end(),
+                    std::back_inserter(resPairs));
+  return resPairs;
+}
+
+//______________________________________________________________________________
+template <typename BinaryPrefilterExpr>
```

Comment:
Why is this templated? I think it always does OR, same for the AND function, maybe an artifact from a previous function where they were mixed?

Reply:
Sorry, this is quite confusing because we kind of only partially apply De-Morgan's law, i will add a more explicit comment that explains this procedure better. The `mergeChildrenForOrExpressionImpl` is templated with `prefilterExpression::OrExpression` or `prefilterExpression::AndExpression` (same for `mergeChildrenForExpressionImpl`), it depends on the value `isNegated`.